docs/Java-SpringBoot-Developer.md

```diff
@@ -0,0 +1,355 @@
+
+This is limitation and needs a spring-boot patch. Spring-Boot entity framework works on entity<-> repo mapping.
+Hence we cannot model movies, genre,actors in the same "movies" collection as recommended in this best practices [document](https://github.com/4-co/imdb)
+spring-data-cosmosdb expects to have 3 separate collections as "movies", "genre", "actor"
+However this is cost implication as "genre" collection having just 30 elements needs a separate collection and is 3 times the RU querying for 3 different collection.
+
+#### Solution
+
+- spring-data-cosmosdb-starter is based on Spring-Boot's Spring-Data framework
+- Spring-Data Defines an entity-specific repository and is built on the following 3 principles:
+  1. Repository pattern - No code repositories
+  2. Reduced boiler plate code for CRUD operations
+  3. Generated Queries ex: findBy
+
+```java
+
+@Document(collection = Constants.DEFAULT_MOVIE_COLLECTION_NAME)
+@JsonPropertyOrder({"id","movieId", "partitionKey",  "type", "title", "textSearch", "year", "runtime", "rating", "votes", "totalScore", "genres", "roles" })
+public class Movie  extends  MovieBase{
+
+  @JsonIgnore
+  @Id
+  private String id;
+
+  @PartitionKey
+  private String partitionKey;
+
+  private double rating;
+  private long votes;
+  private long totalScore;
+  private String textSearch;
+  private List<Role> roles;
+
+
+    public Movie() {
+
+    }
+}
+
+@Repository
+public interface MoviesRepository extends DocumentDbRepository<Movie, String>  {
+    List<Movie> findByMovieId(String movieId);
+    List<Movie> findByTextSearchContaining(String movieName);
+
+}
+
+@Service
+public class MoviesService {
+
+    @Autowired
+    MoviesRepository repository;
+
+    private static Gson gson = new Gson();
+
+
+    public List<Movie> getAllMovies(Optional<String> query, Sort sort) {
+        if (query.isPresent() && !StringUtils.isEmpty(query.get())) {
+            return repository.findByTextSearchContaining(query.get().toLowerCase());
+        } else {
+            return (List<Movie>) repository.findAll(sort);
+        }
+    }
+}
+
+
+```
+
+### Partition Key usage in Spring
+
+### JAVA-SPRINGBOOT-SDK-GAP: spring-boot cosmosdb sdk does not support single document read with partition key​ as it internally calls QueryDocument and not a ReadDocument hence it is not a 1 RU operation
+
+To query by partition id annotate field partition column with @partitionkey in the [document entity](https://github.com/jyotsnaravikumar/helium-java/blob/CSE-feedbacks/src/main/java/com/microsoft/azure/helium/app/movie/Movie.java)
+
+```java
+
+@Document(collection = Constants.DEFAULT_MOVIE_COLLECTION_NAME)
+@JsonPropertyOrder({"id","movieId", "partitionKey",  "type", "title", "textSearch", "year", "runtime", "rating", "votes", "totalScore", "genres", "roles" })
+public class Movie  extends  MovieBase{
+
+  @JsonIgnore
+  @Id
+  private String id;
+
+  @PartitionKey
+  private String partitionKey;
```

Comment:
If spring-boot doesn't use partition key, why do we need it?

Reply:
https://github.com/microsoft/spring-data-cosmosdb/blob/master/QueryPartitionedCollection.md 
We are querying by partition key

Reply:
Does that make it faster / less RUs? If so, makes sense to keep it. If not, does it make sense to have that query be "different"?